crates/polars-core/src/chunked_array/logical/categorical/builder.rs

```diff
@@ -1,465 +1,142 @@
-            _ => false,
-        }
-    }
+// Wrap u32 key to avoid incorrect usage of hashmap with custom lookup
+#[repr(transparent)]
+struct KeyWrapper(u32);
 
-    /// [`str`] to [`Categorical`]
-    ///
-    ///
-    /// [`Categorical`]: crate::datatypes::DataType::Categorical
-    pub fn find(&self, value: &str) -> Option<u32> {
-        match self {
-            Self::Global(rev_map, a, id) => {
-                // fast path is check
-                if using_string_cache() {
-                    let map = crate::STRING_CACHE.read_map();
-                    if map.uuid == *id {
-                        return map.get_cat(value);
-                    }
-                }
-                rev_map
-                    .iter()
-                    // Safety:
-                    // value is always within bounds
-                    .find(|(_k, &v)| (unsafe { a.value_unchecked(v as usize) } == value))
-                    .map(|(k, _v)| *k)
-            },
-
-            Self::Local(a, _) | Self::Enum(a, _) => {
-                // Safety: within bounds
-                unsafe { (0..a.len()).find(|idx| a.value_unchecked(*idx) == value) }
-                    .map(|idx| idx as u32)
-            },
-        }
-    }
-}
-
-#[derive(Eq, Copy, Clone)]
-pub struct StrHashLocal<'a> {
-    str: &'a str,
-    hash: u64,
-}
-
-impl<'a> Hash for StrHashLocal<'a> {
-    fn hash<H: Hasher>(&self, state: &mut H) {
-        state.write_u64(self.hash)
-    }
-}
-
-impl<'a> StrHashLocal<'a> {
-    #[inline]
-    pub(crate) fn new(s: &'a str, hash: u64) -> Self {
-        Self { str: s, hash }
-    }
-}
-
-impl<'a> PartialEq for StrHashLocal<'a> {
-    fn eq(&self, other: &Self) -> bool {
-        // can be collisions in the hashtable even though the hashes are equal
-        // e.g. hashtable hash = hash % n_slots
-        (self.hash == other.hash) && (self.str == other.str)
-    }
-}
-
-pub struct CategoricalChunkedBuilder<'a> {
+pub struct CategoricalChunkedBuilder {
     cat_builder: UInt32Vec,
     name: String,
     ordering: CategoricalOrdering,
-    reverse_mapping: RevMappingBuilder,
+    categories: MutableUtf8Array<i64>,
     // hashmap utilized by the local builder
-    local_mapping: PlHashMap<StrHashLocal<'a>, u32>,
-    // stored hashes from local builder
-    hashes: Vec<u64>,
+    local_mapping: PlHashMap<KeyWrapper, ()>,
 }
 
-impl CategoricalChunkedBuilder<'_> {
+impl CategoricalChunkedBuilder {
     pub fn new(name: &str, capacity: usize, ordering: CategoricalOrdering) -> Self {
-        let builder = MutableUtf8Array::<i64>::with_capacity(capacity / 10);
-        let reverse_mapping = RevMappingBuilder::Local(builder);
-
         Self {
             cat_builder: UInt32Vec::with_capacity(capacity),
             name: name.to_string(),
             ordering,
-            reverse_mapping,
-            local_mapping: Default::default(),
-            hashes: vec![],
+            categories: MutableUtf8Array::<i64>::with_capacity(capacity / 10),
```

Comment:
I know this is copy pasted, but now we are here. Maybe use `HASH_MAP_INITSIZE` here. We found that overallocating hashmaps can be slower due to page faults.

I think we should accept a `cardinality: Option<usize>` argument.